epochStart/metachain/systemSCs.go

```diff
@@ -70,33 +73,79 @@ func NewSystemSCProcessor(args ArgsNewEpochStartSystemSCProcessing) (*systemSCPr
 	}
 
 	return &systemSCProcessor{
-		systemVM:                args.SystemVM,
-		userAccountsDB:          args.UserAccountsDB,
-		peerAccountsDB:          args.PeerAccountsDB,
-		marshalizer:             args.Marshalizer,
-		startRating:             args.StartRating,
-		validatorInfoCreator:    args.ValidatorInfoCreator,
-		endOfEpochCallerAddress: args.EndOfEpochCallerAddress,
-		stakingSCAddress:        args.StakingSCAddress,
-		chanceComputer:          args.ChanceComputer,
+		systemVM:                 args.SystemVM,
+		userAccountsDB:           args.UserAccountsDB,
+		peerAccountsDB:           args.PeerAccountsDB,
+		marshalizer:              args.Marshalizer,
+		startRating:              args.StartRating,
+		validatorInfoCreator:     args.ValidatorInfoCreator,
+		endOfEpochCallerAddress:  args.EndOfEpochCallerAddress,
+		stakingSCAddress:         args.StakingSCAddress,
+		chanceComputer:           args.ChanceComputer,
+		mapNumSwitchedPerShard:   make(map[uint32]uint32),
+		mapNumSwitchablePerShard: make(map[uint32]uint32),
 	}, nil
 }
 
 // ProcessSystemSmartContract does all the processing at end of epoch in case of system smart contract
-func (s *systemSCProcessor) ProcessSystemSmartContract(validatorInfos map[uint32][]*state.ValidatorInfo, epoch uint32) error {
-	err := s.swapJailedWithWaiting(validatorInfos, epoch)
+func (s *systemSCProcessor) ProcessSystemSmartContract(validatorInfos map[uint32][]*state.ValidatorInfo) error {
+	err := s.computeNumWaitingPerShard(validatorInfos)
+	if err != nil {
+		return err
+	}
+
+	err = s.swapJailedWithWaiting(validatorInfos)
 	if err != nil {
 		return err
 	}
+
 	return nil
 }
 
-func (s *systemSCProcessor) auctionSelection() error {
-
+func (s *systemSCProcessor) computeNumWaitingPerShard(validatorInfos map[uint32][]*state.ValidatorInfo) error {
+	for shardID, validatorInfoList := range validatorInfos {
+		totalInWaiting := uint32(0)
+		for _, validatorInfo := range validatorInfoList {
+			switch validatorInfo.List {
+			case string(core.WaitingList):
+				totalInWaiting++
+			case string(core.JailedList):
+				isStaked, err := s.isStaked(validatorInfo.PublicKey)
+				if err != nil {
+					return err
+				}
+				if isStaked {
+					totalInWaiting++
```

Comment:
If you add the jailed as "waitingList" could you get to a point and switch a jailed with another jailed?

Reply:
jailed and staked. thus they were not switched out before. It is not possible to switch one jailed with another one.

Reply:
changed again. nodes with status jailed can be switched. nodes which are validators will be counted to switch only waiting list number.

Reply:
ok